qubes/devices.py

```diff
@@ -115,87 +142,113 @@ class DeviceCollection(object):
     def __init__(self, vm, class_):
         self._vm = vm
         self._class = class_
-        self._set = set()
+        self._set = PersistentCollection()
 
         self.devclass = qubes.utils.get_entry_point_one(
             'qubes.devices', self._class)
 
-    def attach(self, device, persistent=True):
+    def attach(self, device_assignment: DeviceAssignment):
         '''Attach (add) device to domain.
 
         :param DeviceInfo device: device object
         '''
 
-        if device in self.attached():
+        if not device_assignment.frontend_domain:
+            device_assignment.frontend_domain = self._vm
+
+        if device_assignment.frontend_domain != self._vm:
+            raise WrongAssignment(
+                "Trying to attach DeviceAssignment belonging to other domain")
+        if not device_assignment.persistent and self._vm.is_halted():
+            raise WrongAssignment("Devices can only be attached persistent to "
+                                  "a halted vm")
+        device = self._device(device_assignment)
+        if device in self.assignments():
             raise DeviceAlreadyAttached(
-                'device {!r} of class {} already attached to {!r}'.format(
+                'device {!s} of class {} already attached to {!s}'.format(
                     device, self._class, self._vm))
         self._vm.fire_event_pre('device-pre-attach:'+self._class, device=device)
-        if persistent:
-            self._set.add(device)
+        if device_assignment.persistent:
+            self._set.add(device_assignment)
         self._vm.fire_event('device-attach:' + self._class, device=device)
 
-
-    def detach(self, device, persistent=True):
+    def detach(self, device_assignment: DeviceAssignment):
         '''Detach (remove) device from domain.
 
         :param DeviceInfo device: device object
         '''
 
-        if device not in self.attached():
+        if not device_assignment.frontend_domain:
+            device_assignment.frontend_domain = self._vm
+
+        if device_assignment in self._set and not self._vm.is_halted():
+            raise WrongAssignment(
+                "Can not remove a persistent attachment from a non halted vm")
```

Comment:
AFAIR we agreed upon it. I think our motivation was to not being able to remove a devices which is needed for vm functioning. 

I don't mind to allow it.